fix(storage): update instructions for adding external S3 buckets to Amplify (#8289)
aws-amplify/docs: src/pages/[platform]/build-a-backend/storage/use-with-custom-s3/index.mdx

![Showing Amplify console showing Storage tab selected](/images/gen2/storage/s3-console-permissions.png)

The policy will look something like this:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Statement1",
      "Principal": { "AWS": "arn:aws:iam::<AWS-account-ID>:role/<role-name>" },
      "Effect": "Allow",
      "Action": [
        "s3:PutObject",
        "s3:GetObject",
        "s3:DeleteObject",
        "s3:ListBucket"
      ],
      "Resource": [

>         "arn:aws:s3:::<bucket-name>/",

Reviewer: why was this needed in addition?

Reviewer: I dont think you need even that. Why are you changing this block at all? The reason only /* version is need is because Amplify permission only allows for object level not bucket level operations

        "arn:aws:s3:::<bucket-name>/*"
      ]
    }
  ]
}
```
Replace `<AWS-account-ID>` with your AWS account ID and `<role-name>` with the IAM role associated with your Amplify Auth setup. Replace `<bucket-name>` with the S3 bucket name.

You can refer to [Amazon S3's Policies and Permissions documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-policy-language-overview.html) for more ways to customize access to the bucket.

<Callout warning>

> In order to make calls to your S3 bucket from your application, you must also set up a CORS Policy for your S3 bucket. This applies only to manually-configured S3 buckets. Learn more about [setting up a CORS Policy for your S3 bucket](/[platform]/build-a-backend/storage/extend-s3-resources/#for-manually-configured-s3-resources). 

Reviewer: ```suggestion
In order to make calls to your manually configured S3 bucket from your application, you must also set up a [CORS Policy](/[platform]/build-a-backend/storage/extend-s3-resources/#for-manually-configured-s3-resources) for the bucket.
```

</Callout>

### Specify S3 bucket in Amplify's backend config

> Next, use the `addOutput` method from the backend definition object to define a custom S3 bucket by specifying the name and region of the bucket in your **amplify/backend.ts** file. More options can be specified if more granular control over your custom S3 bucket is needed. 

Reviewer: > More options can be specified if more granular control over your custom S3 bucket is needed. 

Dont think this needs explicit callout like this here

          bucket prefixes and configure user access types to them
        */ 
        paths: {
          "public/*": {
            // "write" and "delete" can also be added depending on your use case
            guest: ["get", "list"], 
          },
        },
      }
    ]
  },
});

/*
  Define an inline policy to attach to Amplify's unauth role
  This policy defines how unauthenticated/guest users can access your existing bucket
*/ 
const unauthPolicy = new Policy(backend.stack, "customBucketUnauthPolicy", {
  statements: [
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["s3:GetObject"],
      resources: [`${customBucket.bucketArn}/public/*`],
    }),
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["s3:ListBucket"],
      resources: [
        `${customBucket.bucketArn}`,
        `${customBucket.bucketArn}/*`
      ],
      conditions: {
        StringLike: {
          "s3:prefix": ["public/", "public/*"],
        },
      },
    }),
  ],
});

// Add the policies to the unauthenticated user role
backend.auth.resources.unauthenticatedUserIamRole.attachInlinePolicy(
  unauthPolicy,
);
// highlight-end
```
</Block>

> <Block name="Authenticated Users">

Reviewer 2: can we separate these tabs into individual sections on the page? with tabs it's tough to deep-link to content directly

Reviewer: Also content is mostly repeated i think block makes more sense here right?

Reviewer 2: > With the examples being extremely lengthy, would it be better to instead create a heading right above the examples block for easier linking to that section?

yes, that'd be perfect

> Also content is mostly repeated i think block makes more sense here right?

I think it's fine if we show the diffs between them if the full code snippets are repeated

<Callout>

**Important:** You cannot use both a storage backend configured through Amplify and a custom S3 bucket at the same time. 

If you specify a custom S3 bucket, no sandbox storage resource will be created. The provided custom S3 bucket will be used, even in the sandbox environment.

</Callout>

Below are several examples of configuring the backend to define a custom S3 bucket:

<BlockSwitcher>
<Block name="Guest Users">
Below is an example of expanding the original backend object to grant all guest (i.e. not signed in) users read access to files under `public/`:

```ts title="amplify/backend.ts"
import { defineBackend } from '@aws-amplify/backend';
import { Effect, Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { auth } from "./auth/resource";

const backend = defineBackend({
  auth,
});
// highlight-start
const customBucketStack = backend.createStack("custom-bucket-stack");

// Import existing bucket
const customBucket = Bucket.fromBucketAttributes(bucketStack, "MyCustomBucket", {
  bucketArn: "arn:aws:s3:::<bucket-name>",

>   region: "<region>"

Reviewer: Wanted to call out, Amplify Storage Browser in specific will need the buckets to be in same region because of some restriction in IAM IDC and AccessGrants setup. This is only for managed auth use cases though. That said Amplify Storage APIs itself doesnt bother much about region so i guess choosing one over another.

@josefaidt is there any backend best practices that you are aware of for choosing one over an another.

You can refer to [Amazon S3's Policies and Permissions documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-policy-language-overview.html) for more ways to customize access to the bucket.

<Callout warning>
In order to make calls to your manually configured S3 bucket from your application, you must also set up a [CORS Policy](/[platform]/build-a-backend/storage/extend-s3-resources/#for-manually-configured-s3-resources) for the bucket. 
</Callout>

### Specify S3 bucket in Amplify's backend config

Next, use the `addOutput` method from the backend definition object to define a custom S3 bucket by specifying the name and region of the bucket in your `amplify/backend.ts` file. You must also set up the appropriate resources and IAM policies to be attached to the backend.

<Callout>

> **Important:** You cannot use both a storage backend configured through Amplify and a custom S3 bucket at the same time. 

Reviewer: Can u verify if this is still true?

            authenticated: ["get", "list", "write", "delete"],
            // highlight-end
          },
        },
      }
    ]
  },
});

// ... Unauthenticated/guest user policies and role attachments go here ...
// highlight-start
/*
  Define an inline policy to attach to Amplify's auth role
  This policy defines how authenticated users can access your existing bucket
*/ 
const authPolicy = new Policy(backend.stack, "customBucketAuthPolicy", {
  statements: [
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        "s3:GetObject",
        "s3:PutObject", 
        "s3:DeleteObject"
      ],
      resources: [`${customBucket.bucketArn}/public/*`,],
    }),
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["s3:ListBucket"],
      resources: [
        `${customBucket.bucketArn}`,
        `${customBucket.bucketArn}/*`
        ],
      conditions: {
        StringLike: {
          "s3:prefix": ["public/*", "public/"],
        },
      },
    }),
  ],
});

// Add the policies to the authenticated user role
backend.auth.resources.authenticatedUserIamRole.attachInlinePolicy(authPolicy);
// highlight-end
```
</Block>
<Block name="User Groups">

> Below is an example of expanding the original backend object to have an `admin/` folder that authenticated users can read, but only users belonging to the "admin" user group can manage:

Reviewer: ```suggestion
Below is an example of expanding the original backend object with user group permissions. Here, any authenticated users can read from `admin/` and `public/`and authenticated users belonging to the "admin" user group can only manage `admin/`
```

</Callout>

Below is an example of configuring the backend to define a custom S3 bucket where only authenticated (i.e. signed in) users have full access to a folder called `public/`:

> ```ts title="amplify/backend.ts"

Reviewer: I would merge the sample code into one block switcher since you are already using it and title this part as `Authenticated Users`

import { defineBackend } from "@aws-amplify/backend";
import { auth } from "./auth/resource";

const backend = defineBackend({
  auth,
});
//highlight-start
backend.addOutput({
  storage: {
    aws_region: "<region>",
    bucket_name: "<bucket-name>",
    // optional: `buckets` can be used when setting up more than one existing bucket
    buckets: [
      {
        aws_region: "<region>",
        bucket_name: "<bucket-name>",
        name: "<bucket-name>",

>         // optional: `paths` can be used to set up access to specific bucket prefixes,

Reviewer: nit: use multiline comment

        // and configure user access types to those prefixes
        paths: {
          "public/*": {
            authenticated: ["get", "list", "write", "delete"],
          },
        },
      }
    ]
  },
});

// Define an inline policy to attach to Amplify's auth role

> // This policy defines how authenticated users can access your existing bucket

Reviewer: nit: multiline comment. Check other places to do the same and check to avoid horizontal scrolling because of comments and wrap them up

const authPolicy = new Policy(backend.stack, "customBucketAuthPolicy", {
  statements: [
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["s3:GetObject", "s3:PutObject", "s3:DeleteObject"],
      resources: ["arn:aws:s3:::<bucket-name>/public/*",],
    }),
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["s3:ListBucket"],
      resources: [
        "arn:aws:s3:::<bucket-name>",
        "arn:aws:s3:::<bucket-name>/*"
        ],
      conditions: {
        StringLike: {
          "s3:prefix": ["public/*", "public/"],
        },
      },
    }),
  ],
});

// Add the policies to the authenticated user role
backend.auth.resources.authenticatedUserIamRole.attachInlinePolicy(authPolicy);
// highlight-end
```

From there, you can further configure the backend to add custom authorization rules for different user types. 

> <Callout>

Reviewer: move the call out to below the code block

The custom authorization rules defined in the examples below are able to be combined and follow the same rules used when working with Amplify-defined storage. For more information about the access types and access definition rules supported by Amplify, please refer to our documentation on [customizing authorization rules](/[platform]/build-a-backend/storage/authorization/). 
</Callout>

<BlockSwitcher>
<Block name="Guest Users">
Below is an example of expanding the original backend object to grant all guest (i.e. not signed in) users read access to files under `public/`:

```ts title="amplify/backend.ts"

// ...

backend.addOutput({
  storage: {

>     aws_region: "<region>",

Reviewer 2: can we pivot this to `s3.Bucket.fromBucketName()`? this should help with the access control shown later https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3.Bucket.html#static-fromwbrbucketwbrnamescope-id-bucketname

      actions: ["s3:ListBucket"],
      resources: [
        "arn:aws:s3:::<bucket-name>",
        "arn:aws:s3:::<bucket-name>/*"
      ],
      conditions: {
        StringLike: {
          "s3:prefix": ["public/", "public/*"],
        },
      },
    }),
  ],
});

// Add the policies to the unauthenticated user role
backend.auth.resources.unauthenticatedUserIamRole.attachInlinePolicy(
  unauthPolicy,
);
// highlight-end
```
</Block>
<Block name="User Groups">
Below is an example of expanding the original backend object to have an `admin/` folder that only users belonging to the "admin" user group can manage:
{/* cSpell:disable */}
```ts title="amplify/backend.ts"

// ...

backend.addOutput({
  storage: {
    aws_region: "<region>",
    bucket_name: "<bucket-name>",
    buckets: [
      {
        aws_region: "<region>",
        bucket_name: "<bucket-name>",
        name: "<bucket-name>",
        // @ts-expect-error: Amplify backend type issue
        // https://github.com/aws-amplify/amplify-backend/issues/2569
        paths: {
          "public/*": {
            authenticated: ["get", "list", "write", "delete"],
          },
          // highlight-start
          "admin/*": {

>             groupsadmin: ["get", "list", "write", "delete"],

Reviewer: i think customers will also benefit from not admin users accessing this. Like you can `authenticated` to just `read` scope and explain on top

          },
          // highlight-end
        },
      }
    ]
  },
});

// ... Authenticated user policy and role attachment goes here ...

// highlight-start
// Define an inline policy to attach to "admin" user group role
// This policy defines how authenticated users with "admin" user group role can access your existing bucket
const adminPolicy = new Policy(backend.stack, "customBucketAdminPolicy", {
  statements: [
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["s3:GetObject", "s3:PutObject", "s3:DeleteObject"],
      resources: ["arn:aws:s3:::<bucket-name>/admin/*"],
    }),
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["s3:ListBucket"],
      resources: [
        "arn:aws:s3:::<bucket-name>",
        "arn:aws:s3:::<bucket-name>/*",
      ],
      conditions: {
        StringLike: {
          "s3:prefix": ["admin/*", "admin/"],
        },
      },
    }),
  ],
});

// Add the policies to the "admin" user group role
backend.auth.resources.groups["admin"].role.attachInlinePolicy(adminPolicy);
// highlight-end
```
{/* cSpell:enable */}
</Block>
<Block name="Owners">

> While Amplify Storage uses the reserved token `entity_id` to represent the user's identity ID, that token is not available when configuring external buckets. Instead, you must use [the IAM role `${cognito-identity.amazonaws.com:sub}`](https://docs.aws.amazon.com/cognito/latest/developerguide/iam-roles.html#trust-policies-classic) to specify the user's identity ID.

Reviewer: i think we can rephrase this to just callout about `{cognito-identity.amazonaws.com:sub}` for using identityId for the current owner only access

{/* cSpell:enable */}
</Block>
<Block name="Owners">

> Amplify allows scoping file access to individual users via the user's identity ID. To specify the user's identity ID, you can use the token `${cognito-identity.amazonaws.com:sub}`.

Reviewer: nit: i would merge this with the para below and follow same pattern as other blocks to start with `Below is an`

import { Bucket } from 'aws-cdk-lib/aws-s3';
import { auth } from "./auth/resource";

const backend = defineBackend({
  auth,
});

const customBucketStack = backend.createStack("custom-bucket-stack");

// Import existing bucket
const customBucket = s3.Bucket.fromBucketAttributes(bucketStack, "MyCustomBucket", {
  bucketArn: "arn:aws:s3:::<bucket-name>",
  region: "<region>"
});

backend.addOutput({
  storage: {
    aws_region: customBucket.env.region,
    bucket_name: customBucket.bucketName,
    buckets: [
      {
        aws_region: customBucket.env.region,
        bucket_name: customBucket.bucketName,
        name: customBucket.bucketName,
        /*
          @ts-expect-error: Amplify backend type issue
          https://github.com/aws-amplify/amplify-backend/issues/2569
        */ 
        paths: {
          "public/*": {
            guest: ["get", "list"],
            authenticated: ["get", "list", "write", "delete"],
          },
          // highlight-start
          // allow all users to view all folders/files within `protected/`
          "protected/*": {
            guest: ["get", "list"],
            authenticated: ["get", "list"],
          },

>           // allow owners to get/modify/delete their own files in assigned subfolder

Reviewer: ```suggestion
          // allow owners to read and write their own files in assigned subfolder
```

### Using Amplify configure
Existing storage resource setup can be accomplished by passing the resource metadata to `Amplify.configure`. This will configure the Amplify Storage client library to interact with the additional resources. It's recommended to add the Amplify configuration step as early as possible in the application lifecycle, ideally at the root entry point.

{/* cSpell:disable */}
```ts
import { Amplify } from "aws-amplify";

Amplify.configure({
  Auth: {
    // add your auth configuration
  },
  Storage: {
    S3: {
      bucket: "<your-default-bucket-name>",
      region: "<your-default-bucket-region>",
      // default bucket metadata should be duplicated below with any additional buckets
      buckets: {
        "<your-default-bucket-friendly-name>": {
          bucketName: "<your-default-bucket-name>",
          region: "<your-default-bucket-region>",
          paths: {

>             "public/*": {

Reviewer: did we test this to ensure the working of `listPaths` is alright?

{/* cSpell:enable */}
</Block>
</BlockSwitcher>

<Callout>

> The custom authorization rules defined in the examples are able to be combined, and follow the same rules used when working with Amplify-defined storage. For more information about the access types and access definition rules supported by Amplify, please refer to our documentation on [customizing authorization rules](/[platform]/build-a-backend/storage/authorization/). 

Reviewer: ```suggestion
The custom authorization rules defined in the examples can be combined, and follow the same rules as Amplify-defined storage. Please refer to our documentation on [customizing authorization rules](/[platform]/build-a-backend/storage/authorization/) for more information.
```
